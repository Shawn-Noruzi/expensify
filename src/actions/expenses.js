import database from "../firebase/firebase";
// If your action is doing something asynchronous like accessing a database, then redux-thunk expects your action to return a function. redux-thunk then passes that function the dispatch argument which can be called once the async operation is complete to dispatch the action to your reducers.

// ADD_EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database
      .ref("expenses")
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

//async function so we use return inside -> then calls another function (ie. removeExpense)
export const startRemoveExpense = ({ id } = {}) => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

//SET_EXPENSES - sets array value, gets it back from DB
export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

//This is our asynch action -> fetches then dispatches setExpenses
//no args needed since its just fetching.
//1. fetch all expense data once
//2.parse data into an array
//3.Dispatch SET_EXPENSES

export const startSetExpenses = () => {
  return dispatch => {
    return database
      .ref("expenses")
      .once("value")
      .then(snapshot => {
        const expenses = [];

        snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });

        dispatch(setExpenses(expenses));
      });
  };
};
