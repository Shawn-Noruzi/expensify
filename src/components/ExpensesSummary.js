import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expenses-total";

//expenseCount - how many visisble expenses
//expensesTotal - total of visible expenses?
//2 snapshot tests
//sfc used here because we dont want this component to be able to
//change the state.

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? "expense" : "expenses";
  const formattedExpensesTotal = numeral(expensesTotal / 100).format("$0,0.00");
  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}
      </h1>
    </div>
  );
};

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
