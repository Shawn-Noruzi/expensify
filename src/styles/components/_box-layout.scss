.box-layout {
  //set a BG image in /public/images that covers the whole page
  background: url("/images/bg.jpg");
  background-size: cover;

  //positioning of elements in the center of the page
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  //setting the size of the entire div -> 100% of the viewport height and width
  height: 100vh;
  width: 100vw;
}

.box-layout__box {
  background: fade-out($color: rgb(255, 255, 255), $amount: 0.5);
  border-radius: 4px;
  text-align: center;
  width: 30rem;
  margin: $m-size;
  padding: $m-size;
}

.box-layout__title {
  margin: $s-size 0 0 0 ;
  line-height: 1;
  padding: $m-size;
  color: white;
}

.box-layout__paragraph {
  margin: $s-size 0 0 0 ;
  line-height: 1;
  padding: $s-size;
  font-size: 2 rem;
  color: white;
  text-align: start;
}

