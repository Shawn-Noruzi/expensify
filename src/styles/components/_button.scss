.button {
    background: rgb(22, 145, 182);
    color: white;
    border-radius: 4px;
    border: none;
    text-align: center;
    padding: 20px;
    transition: all 0.5s;
    cursor: pointer;
    margin-top: $s-size;
    margin-bottom: $l-size;
    font-size: $m-size;
  }
  .button span {
    cursor: pointer;
    display: inline-block;
    position: relative;
    transition: 0.5s;
  }
  
  .button span:after {
    content: "\00bb";
    position: absolute;
    opacity: 0;
    top: 0;
    right: -20px;
    transition: 0.5s;
  }
  
  .button:hover span {
    padding-right: 25px;
  }
  
  .button:hover span:after {
    opacity: 1;
    right: 0;
  }
  