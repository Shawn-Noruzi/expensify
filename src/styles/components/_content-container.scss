.content-container{
    margin: 0 1.8rem !important;
    padding: 0 $m-size;
    max-width:80rem;
}