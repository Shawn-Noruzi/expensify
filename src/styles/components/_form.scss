.form {
    display: flex;
    flex-direction: column;
    //selection for children of a tag ie.items under form
    > * {
        margin-bottom: $m-size;
    }
}



.form__error {
    margin: 0 0 $m-size 0;
    font-style: italic;
}