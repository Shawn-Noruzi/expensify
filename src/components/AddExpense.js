import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";

//to avoid inline functions we're switching to class component

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.addExpense(expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h1> Add Expense </h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
});
export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);

