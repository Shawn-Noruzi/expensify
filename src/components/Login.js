import React from "react";
import { connect } from "react-redux";
import { startLogin } from "../actions/auth";

//doesnt need access to state -> dont need to provide stateToProps
//needs dispatchToProps -> bc we want to dispatch an action (startLogin)

export const LoginPage = ({ startLogin }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Expensify</h1>
      <button className="button" onClick={startLogin}><span>Login</span></button>
    </div>
  </div>
);

const mapDispatchToProps = dispatch => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(
  undefined,
  mapDispatchToProps
)(LoginPage);
