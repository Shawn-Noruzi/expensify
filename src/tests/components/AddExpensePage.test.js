import React from "react";
import { shallow } from "enzyme";
import { AddExpensePage } from "../../components/AddExpense";
import expenses from "../fixtures/expenses";

let addExpense, history, wrapper;
beforeEach(() => {
  addExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <AddExpensePage addExpense={addExpense} history={history} />
  );
});

test("should add expense correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test(" should handle onSubmit ", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(addExpense).toHaveBeenLastCalledWith(expenses[1]);
});
