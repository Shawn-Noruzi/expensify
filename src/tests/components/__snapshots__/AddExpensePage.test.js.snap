// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add expense correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1>
      Add Expense
    </h1>
  </div>
  <div />
  <div
    className="content-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
